### PR TITLE
Removal of browser-storage-memory

### LINKED ISSUE #11173: Remove browser-storage-memory
This component was created mostly for testing purposes as a simple implementation of the `concept-storage`. While it was useful in the early days of A-C as a way to test-fit our storage components into things, we haven't really kept it up, and are not getting much value out of this code. It's time to remove it!

### PR DESCRIPTION
**Pull request by 4shutosh**
### Pull Request checklist
Fixes #11173 

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by 4shutosh** [comment]
I was not sure about the tests using the `InMemoryHistoryStorage` so I've created a class for each of them.
Please let me know if that's wrong and what other changes are required.

**Review by pocmo — COMMENTED** [review]
Thank you for your PR! The removal looks good.

I'll defer to @grigoryk regarding the unit tests. The duplication of the in-memory implementation is a bit unfortunate.

File: components/concept/storage/README.md [hunk 1]
```diff
@@ -5,7 +5,6 @@ The `concept-storage` component contains interfaces and abstract classes that de
 This abstraction makes it possible to build components that work independently of the storage layer being used.
 
 Currently two store implementations are available:
```
Comment (grigoryk) [inline components/concept/storage/README.md:7]: Can you update this comment as well? "Currently a single store implementation is available:"

**Review by grigoryk — COMMENTED** [review]
The deletion looks good, thanks!

As for the tests, I don't think we need to duplicate these fake implementations.

I think if we were writing these tests from scratch now, we'd just use a mock of the storage, and configure that mock to return the data necessary for the specific tests in question. This will involve updating all of these tests - whenever we add something to the storage, provide the data via the mock instead.

This matches the "test pyramid" kind of approach - we're not testing the storage here, just how we interact with it.

Another option which is also fine and much easier is to update these tests to use `PlacesHistoryStorage` as a quick way to move forward here. The main downside is that it'll slow down the tests a bit, I think - but probably not by too much.

You'd need to add a testImplementation of `browser-storage-sync` and `testImplementation Dependencies.mozilla_full_megazord_forUnitTests` to the gradle files for the affected modules.
Then, you can just just swap out the storage instances in tests with something like `val historyStore = PlacesHistoryStorage(testContext)`.

For an example of this, tests in `support-migration` use the `browser-storage-sync` storage layer.

File: components/concept/storage/README.md (hunk 1 above)
Comment (4shutosh) [reply components/concept/storage/README.md:7]: Ohh yes, such a silly one. Will do that!

**Comment by 4shutosh** [comment]
Thank you for the detailed explanation @grigoryk.

I've done the changes suggested by you, please take a look and let me know if any other changes are required.

**Comment by pocmo** [comment]
There are 11 failed tests in `feature-awesomebar` and 1 failed test in `feature-toolbar`. Does this reproduce locally?

**Comment by 4shutosh** [comment]
Yes, they failed locally as well.
image
the [log](https://firefox-ci-tc.services.mozilla.com/tasks/OS8Y5ZmaQWe-prOxgXV56w/runs/0/logs/public/logs/live.log) is similar to the online one

**Comment by 4shutosh** [comment]
Now the tests are getting instrumented correctly but are failing logically due to the usage of PlacesHistoryStorage.

I'll have to dig deep to get an understanding of how they are working. Any other option to work with :)?

**Comment by grigoryk** [comment]
> Now the tests are getting instrumented correctly but are failing logically due to the usage of PlacesHistoryStorage.
> 
> I'll have to dig deep to get an understanding of how they are working. Any other option to work with :)?

Thanks for updating the PR! It seems like assumptions tests were making about behaviour of the old in-memory store don't apply to the real one. I think it's a good thing actually, maybe we'll uncover some real bugs here!

I think the way forward is to work through each test failure one-by-one, seeing what's going wrong, fixing the tests and maybe fixing underlying issue if there is one. That's likely quite a bit past the `good-first-issue` level of involvement at this point, so I'm happy to work through it locally. I'll start with the `awesomebar` failures, feel free to push more commits to the PR as well.

**Comment by 4shutosh** [comment]
Okay. I'll try to work with the `toolbar` failures then.

**Comment by grigoryk** [comment]
@4shutosh I've updated tests in `feature-awesomebar`, let's see if they pass now. Take a look - I've removed the `browser-storage-sync` dependency entirely, switching tests to use mocked storage.

It'd be great if you could do the same `feature-toolbar` while we're at it, switch it to use mocked storage (e.g. following the pattern in my commit), and remove the unnecessary `testImplementation` dependencies.

**Comment by 4shutosh** [comment]
Sure. On it.

**Comment by 4shutosh** [comment]
I've done the changes but encountered this failure
image

**Comment by grigoryk** [comment]
> I've done the changes but encountered this failure image

Yeah, that's expected. You switched to use a `mock`, but kept the old `recordVisit` calls - they don't actually do anything anymore, and so the test isn't passing now.

You'll need to replace bits where we insert stuff into the storage with corresponding mock statements.

For example:
```
runBlocking {
            history.recordVisit("https://www.mozilla.org", PageVisit(VisitType.TYPED, RedirectSource.NOT_A_SOURCE))
}
```

should be replaces by something like `doReturn(listOf(SearchResult(...)).`when`(history).getSuggestions(...)`.

See https://github.com/mozilla-mobile/android-components/pull/11211/commits/3af92afe2eded8a6d6bdfc0eb09bcadc2bd8f2a9 for examples.

**Comment by 4shutosh** [comment]
Thank you for the beautiful explanation @grigoryk.

The tests pass now.

I feel I did not try enough to look for that, that was easy enough to be done by myself.

Are there any other things that need to be done for this PR? If not please assign me another issue or task as you know better what next should I get going with on this project.

File: components/feature/awesomebar/build.gradle [hunk 2]
```diff
@@ -42,14 +42,15 @@ dependencies {
...
-    testImplementation project(':browser-storage-memory')
     testImplementation project(':lib-fetch-httpurlconnection')
 
     testImplementation Dependencies.androidx_test_core
     testImplementation Dependencies.androidx_test_junit
     testImplementation Dependencies.testing_robolectric
     testImplementation Dependencies.testing_mockito
     testImplementation Dependencies.testing_mockwebserver
+
+    testImplementation Dependencies.mozilla_glean_forUnitTests
```
Comment (grigoryk) [inline components/feature/awesomebar/build.gradle:53]: I don't think we need this?

File: components/concept/storage/README.md [hunk 3]
```diff
@@ -4,8 +4,7 @@ The `concept-storage` component contains interfaces and abstract classes that de
 
 This abstraction makes it possible to build components that work independently of the storage layer being used.
 
-Currently two store implementations are available:
-- [in-memory storage](../../browser/storage-memory)
+Currently a single store implementations is available:
```
Comment (grigoryk) [inline components/concept/storage/README.md:7]: spelling: replace `implementations` w/ `implementation`.

File: components/feature/toolbar/build.gradle [hunk 4]
```diff
@@ -44,13 +44,14 @@ dependencies {
...
     testImplementation project(':support-test')
-    testImplementation project(':browser-storage-memory')
 
     testImplementation Dependencies.androidx_test_core
     testImplementation Dependencies.androidx_test_junit
     testImplementation Dependencies.testing_mockito
     testImplementation Dependencies.testing_robolectric
     testImplementation Dependencies.testing_coroutines
+
+    testImplementation Dependencies.mozilla_glean_forUnitTests
```
Comment (grigoryk) [inline components/feature/toolbar/build.gradle:54]: Also, don't need this, right?

**Review by grigoryk — COMMENTED** [review]
Just a few small things and this should be good to land!

**Comment by grigoryk** [comment]
> Thank you for the beautiful explanation @grigoryk.
> 
> The tests pass now.
> 
> I feel I did not try enough to look for that, that was easy enough to be done by myself.
> 
> Are there any other things that need to be done for this PR? If not please assign me another issue or task as you know better what next should I get going with on this project.

No worries, and thanks for contributing! Looking at the list of issues, we're a bit short right now, but that doesn't mean there aren't good projects/issues to work on. I'll look around and will post some links here tomorrow.

**Review by grigoryk — APPROVED** [review]
✨

**Comment by 4shutosh** [comment]
Thank you for the help 🙌🏻 

> I'll look around and will post some links here tomorrow.

Yes please do.

**Pull request merged**